src/cdk/tree/nested-node.ts

```diff
@@ -72,14 +72,24 @@ export class CdkNestedTreeNode<T, K = T>
     if (!this._tree.treeControl.getChildren && (typeof ngDevMode === 'undefined' || ngDevMode)) {
       throw getTreeControlFunctionsMissingError();
     }
-    const childrenNodes = this._tree.treeControl.getChildren(this.data);
-    if (Array.isArray(childrenNodes)) {
-      this.updateChildrenNodes(childrenNodes as T[]);
-    } else if (isObservable(childrenNodes)) {
-      childrenNodes
-        .pipe(takeUntil(this._destroyed))
-        .subscribe(result => this.updateChildrenNodes(result));
-    }
+
+    this._dataChanges
+      .pipe(
+        takeUntil(this._destroyed),
+        startWith(undefined as void),
+        switchMap(() => {
+          const childrenNodes = this._tree.treeControl.getChildren(this.data);
+          if (Array.isArray(childrenNodes)) {
+            return of(childrenNodes);
+          } else if (isObservable(childrenNodes)) {
+            return childrenNodes;
+          } else {
+            return EMPTY;
+          }
+        }),
+      )
+      .subscribe(result => this.updateChildrenNodes(result));
```

Comment:
This change makes sure that this directive actually responds to data changes. 

However, this is currently only part of the fix, since `cdkTreeNode.mostRecentTreeNode.data` is only updated when a new node is inserted. I am still looking for a solution in the `CdkTree` that also works with data changes.

For now, my workaround is:
```
<fixed-nested-tree-node *cdkTreeNodeDef="let node; when: hasChild" [node]="node">
```
```ts
export class FixedNestedTreeNode<T, K = T> extends CdkNestedTreeNode<T, K>
{
  @Input()
  set node(value: T) {
    this.data = value;
  }
}
```